File: iosApp/iosApp/Modules/NFTsCollection/NFTsCollectionInteractor.swift
// Created by web3d4v on 29/05/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

protocol NFTsCollectionInteractor: AnyObject {
    func fetchCollection(
        with identifier: String,
        onCompletion: (Result<NFTCollection, Error>) -> Void
    )
    func fetchNFTs(
        forCollection identifier: String,
        onCompletion: (Result<[NFTItem], Error>) -> Void
    )
}

final class DefaultNFTsCollectionInteractor {
    private var service: NFTsService

    init(service: NFTsService) {
        self.service = service
    }
}

extension DefaultNFTsCollectionInteractor: NFTsCollectionInteractor {
    
    func fetchCollection(
        with identifier: String,
        onCompletion: (Result<NFTCollection, Error>) -> Void
    ) {
        onCompletion(.success(service.collection(identifier: identifier)))
    }
    
    func fetchNFTs(
        forCollection identifier: String,
        onCompletion: (Result<[NFTItem], Error>) -> Void
    ) {
        onCompletion(.success(service.yourNFTs(collectionId: identifier)))
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/nftsCollection/NFTsCollectionInteractor.kt
package com.sonsofcrypto.web3walletcore.modules.nftsCollection

import com.sonsofcrypto.web3walletcore.services.nfts.NFTCollection
import com.sonsofcrypto.web3walletcore.services.nfts.NFTItem
import com.sonsofcrypto.web3walletcore.services.nfts.NFTsService

interface NFTsCollectionInteractor {
    fun collection(collectionId: String): NFTCollection
    fun nfts(collectionId: String): List<NFTItem>
    fun clearCache()
}

class DefaultNFTsCollectionInteractor(
    private val nftsService: NFTsService,
): NFTsCollectionInteractor {
    private var collections: MutableMap<String, NFTCollection> = mutableMapOf()
    private var nfts: MutableMap<String, List<NFTItem>> = mutableMapOf()

    override fun clearCache() {
        collections = mutableMapOf()
        nfts = mutableMapOf()
    }

    override fun collection(collectionId: String): NFTCollection {
        val collection = collections[collectionId]
            ?: nftsService.collection(collectionId)
        collections[collectionId] = collection
        return collection

    }


    override fun nfts(collectionId: String): List<NFTItem> {
        val collectionNfts = nfts[collectionId]
            ?: nftsService.nfts(collectionId)
        nfts[collectionId] = collectionNfts
        return collectionNfts
    }
}
